src/model/containers.ts

```diff
@@ -0,0 +1,121 @@
+    private _event_since_load: boolean = false;
+
+    constructor() {
+        const supports_containers = [
+            typeof contextualIdentities?.query,
+            typeof contextualIdentities?.onCreated?.addListener,
+            typeof contextualIdentities?.onUpdated?.addListener,
+            typeof contextualIdentities?.onRemoved?.addListener]
+            .every(v => v === 'function');
+        if (!supports_containers) {
+            // Avoid blowing up if running on a browser with no container support.
+            this.enabled = false;
+            return;
+        }
+
+        this.enabled = true;
+        const wiring = new EventWiring(this, {
+            onFired: () => {this._event_since_load = true;},
+            // istanbul ignore next -- safety net; reload the model in the event
+            // of an unexpected exception.
+            onError: () => {logErrorsFrom(() => this.reload());},
+        });
+
+        wiring.listen(contextualIdentities.onCreated, this.whenChanged);
+        wiring.listen(contextualIdentities.onUpdated, this.whenChanged);
+        wiring.listen(contextualIdentities.onRemoved, this.whenRemoved);
+    }
+
+    static async from_browser(): Promise<Model> {
+        const model = new Model();
+        await model.reload();
+        return model;
+    }
+
+    // Fetch tabs/windows from the browser again and update the model's
+    // understanding of the world with the browser's data.  Use this if it looks
+    // like the model has gotten out of sync with the browser (e.g. for crash
+    // recovery).
+    readonly reload = backingOff(async () => {
+        if (!this.enabled) return;
+        // We loop until we can complete a reload without receiving any
+        // concurrent events from the browser--if we get a concurrent event, we
+        // need to try loading again, since we don't know how the event was
+        // ordered with respect to the query().
+        let loaded_containers: Container[];
+
+        this._event_since_load = true;
+        while (this._event_since_load) {
+            this._event_since_load = false;
+            try {
+                loaded_containers = await contextualIdentities.query({});
+            } catch (e) {
+                // Containers can be manually disabled in Firefox by setting
+                // privacy.userContext.enabled to false. When this is the case,
+                // calling .query() will raise an exception.
+                this.enabled = false;
+                loaded_containers = [];
+                break;
+            }
+        }
+        this.containers = new Map(loaded_containers!
+            .filter(c => c?.cookieStoreId)
+            .map(c => [c.cookieStoreId, c]));
+    });
+
+    // Accessors
+    container(key: string): Container | undefined {
+        // Something like a blank string should never be valid.
+        return key ? this.containers.get(key) : undefined;
```

Comment:
This might be overly pedantic, but I don't know if this comment is actually true? MDN doesn't really say what a valid cookieStoreId is, only that it's a string. So the empty string might well be a valid cookieStoreId (though in practice it looks like your comment is correct).

Reply:
I'm not sure either. I think it was just to indicate that relying on Javascript truthiness there was deliberate. Do you think the logic needs to change also? Right now it doesn't consider blank strings.

Reply:
Yeah, I thought about it a bit more, and I think it's probably safer to assume the empty string might be a valid ID. That way if Firefox does something weird in the future (which seems to happen more often than you'd think...), things are less likely to break.

Reply:
>  I think it's probably safer to assume the empty string might be a valid ID.

Fair enough. I'll change this tomorrow since I'd probably want to look through the rest of the code to make sure there's nothing assuming blank container ids can't exist or inconsistent filtering of them.

Reply:
Fixed. Other code that cared about an empty string or used it as a default also updated.